Add Support for some kubernetes client settings

duyanghao opened the pull request:
## What changes were proposed in this pull request?

Add Support for `spark.kubernetes.client.watch.reconnectInterval`(which is relevant to KUBERNETES_WATCH_RECONNECT_INTERVAL_SYSTEM_PROPERTY in io.fabric8.kubernetes.client.Config.)
Add Support for `spark.kubernetes.client.watch.reconnectLimit`(which is relevant to KUBERNETES_WATCH_RECONNECT_LIMIT_SYSTEM_PROPERTY in io.fabric8.kubernetes.client.Config.)
Add Support for `spark.kubernetes.client.connection.timeout`(which is relevant to KUBERNETES_CONNECTION_TIMEOUT_SYSTEM_PROPERTY in io.fabric8.kubernetes.client.Config.)
Add Support for `spark.kubernetes.client.request.timeout`(which is relevant to KUBERNETES_REQUEST_TIMEOUT_SYSTEM_PROPERTY in io.fabric8.kubernetes.client.Config.)

## How was this patch tested?
start submit with following pars:
```
--conf spark.kubernetes.client.connection.timeout=30000 \
--conf spark.kubernetes.client.request.timeout=30000 \
--conf spark.kubernetes.client.watch.reconnectInterval=5000 \
--conf spark.kubernetes.client.watch.reconnectLimit=5 \
```

Tests on my cluster show that above settings do work.

liyinan926 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/SparkKubernetesClientFactory.scala:
```diff
@@ -81,6 +81,30 @@ private[spark] object SparkKubernetesClientFactory {
         }.withOption(namespace) {
           (ns, configBuilder) => configBuilder.withNamespace(ns)
         }.build()
+    val conn_timeout_properties = sparkConf
```
Camel case please for all the new vals.

duyanghao replied:
Done

liyinan926 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:
```diff
@@ -545,4 +545,36 @@ package object config extends Logging {
   }
+
+  private[spark] val KUBERNETES_CLIENT_WATCH_RECONNECT_INTERVAL_SYSTEM_PROPERTY =
+    ConfigBuilder("spark.kubernetes.client.watch.reconnectInterval")
```
The configuration properties should include time unit in the property names, e.g., `reconnectInternalInMs`.

duyanghao replied:
Done

liyinan926 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:
```diff
@@ -545,4 +545,36 @@ package object config extends Logging {
+
+  private[spark] val KUBERNETES_CLIENT_WATCH_RECONNECT_INTERVAL_SYSTEM_PROPERTY =
+    ConfigBuilder("spark.kubernetes.client.watch.reconnectInterval")
+      .doc("spark kubernetes client watch reconnectInterval," +
```
All of our config properties docs start with a Capital letter.

duyanghao replied:
Done

liyinan926 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:
```diff
@@ -545,4 +545,36 @@ package object config extends Logging {
   }
+
+  private[spark] val KUBERNETES_CLIENT_WATCH_RECONNECT_INTERVAL_SYSTEM_PROPERTY =
+    ConfigBuilder("spark.kubernetes.client.watch.reconnectInterval")
```
You might want to define a command prefix for the keys.

duyanghao replied:
yeah

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:
```diff
@@ -545,4 +545,36 @@ package object config extends Logging {
+      .doc("spark kubernetes client watch reconnectInterval," +
+        "which is relevant to KUBERNETES_WATCH_RECONNECT_INTERVAL_SYSTEM_PROPERTY" +
+        " in io.fabric8.kubernetes.client.Config.")
+      .stringConf
```
Use `intConf` with default values everywhere.

mccheah replied:
Actually `timeConf`might make some sense in some of these cases. I think these should have default values that are specified in the config entries and not in the main logic.

duyanghao replied:
Done

mccheah commented:
Document all new properties in `docs/running-on-kubernetes.md` as well as in https://github.com/apache-spark-on-k8s/userdocs after this merges.

duyanghao commented:
@mccheah 
>Document all new properties in docs/running-on-kubernetes.md as well as in https://github.com/apache-spark-on-k8s/userdocs after this merges.

Done for `docs/running-on-kubernetes.md`

liyinan926 commented:
Please rebase against the latest branch-2.2-kubernetes.

mccheah commented on docs/running-on-kubernetes.md:
```diff
@@ -798,6 +798,38 @@ from the other deployment modes. See the [configuration page](configuration.html
+  <td><code>1s</code></td>
+  <td>
+    Spark kubernetes client watch reconnectInterval, which is relevant to
+    <code>KUBERNETES_WATCH_RECONNECT_INTERVAL_SYSTEM_PROPERTY</code> in io.fabric8.kubernetes.client.Config.
```
Use semantic meanings here as opposed to meanings specific to the code implementation. For example, this documentation becomes incorrect if we end up not using the Fabric8 Kubernetes client. The other settings provide some examples that this can be based on.

duyanghao replied:
Done

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:
```diff
@@ -545,4 +545,36 @@ package object config extends Logging {
+
+  private[spark] val KUBERNETES_CLIENT_WATCH_RECONNECT_LIMIT_SYSTEM_PROPERTY =
+    ConfigBuilder("spark.kubernetes.client.watch.reconnectLimit")
+      .doc("Spark kubernetes client watch reconnectLimit," +
```
Also use semantic meanings in these docs, as opposed to implementation specific. "Limit of times where connections can be attempted for watching resources" or something.

duyanghao replied:
Done

duyanghao commented:
@mccheah @liyinan926 
>Please rebase against the latest branch-2.2-kubernetes.

Done

liyinan926 commented on docs/running-on-kubernetes.md:
```diff
@@ -798,6 +798,34 @@ from the other deployment modes. See the [configuration page](configuration.html
+  <td><code>spark.kubernetes.client.watch.reconnectLimit</code></td>
+  <td><code>-1</code></td>
+  <td>
+    Limit of times where connections can be attempted for kubernetes client request.
```
Delete `where`. And change `request` to `requests`.

duyanghao replied:
Done

liyinan926 commented:
Minor comment. Also it seems this is still out-of-date with the base branch.

duyanghao reviewed (commented):
@liyinan926 rebase Done.

duyanghao committed:
Add Support for some kubernetes client settings

Add Support for `spark.kubernetes.client.watch.reconnectInterval`(which is relevant to KUBERNETES_WATCH_RECONNECT_INTERVAL_SYSTEM_PROPERTY in io.fabric8.kubernetes.client.Config.)
Add Support for `spark.kubernetes.client.watch.reconnectLimit`(which is relevant to KUBERNETES_WATCH_RECONNECT_LIMIT_SYSTEM_PROPERTY in io.fabric8.kubernetes.client.Config.)
Add Support for `spark.kubernetes.client.connection.timeout`(which is relevant to KUBERNETES_CONNECTION_TIMEOUT_SYSTEM_PROPERTY in io.fabric8.kubernetes.client.Config.)
Add Support for `spark.kubernetes.client.request.timeout`(which is relevant to KUBERNETES_REQUEST_TIMEOUT_SYSTEM_PROPERTY in io.fabric8.kubernetes.client.Config.)

Signed-off-by: duyanghao <1294057873@qq.com>

liyinan926 commented:
LGTM.

mccheah committed:
Merge branch 'branch-2.2-kubernetes' into branch-2.2-kubernetes

mccheah commented:
After the build passes following the updated branch we can merge this.

duyanghao committed:
rebase

Signed-off-by: duyanghao <1294057873@qq.com>

duyanghao committed:
Add Recovery Logic for Failed Pod

Signed-off-by: duyanghao <1294057873@qq.com>
